### PR TITLE
ceph: Only raise CephCluster conditions that represent current state

### PR DESCRIPTION
**Pull request by travisn**
**Description of your changes:**
The conditions on the cephcluster CR were being set to false when not in progress, which is misleading to the purpose of conditions. The conditions will now always represent current state of the cluster. Transient conditions that show progress of a reconcile will be removed from the status.

Still working out the kinks, not quite ready for v.1.5.8...

**Which issue is resolved by this Pull Request:**
Resolves #7262 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Review by BlaineEXE — COMMENTED** [review]
To clarify for my understanding:
- some conditions are transient ("Connecting", "Progressing", "Failure", "Deleting")
- some conditions are persistent ("Connected", "Ready")
- transient conditions are added during reconcile
- transient conditions are removed when a persistent condition is set ("Connected" or "Ready" is set/reset after reconciling is done)

Questions
- should "Failure" conditions be persistent for errors encountered during provisioning? I'm still trying to figure out how we can do things like present errors to the user while still continuing with provisioning. Like if an OSD Deployment can't be updated, for example. We can still claim the cluster is "Created" (I think), but we also have an error.

**Comment by BlaineEXE** [comment]
Should we document the conditions and their behavior?

File: pkg/operator/ceph/cluster/cephstatus.go
Comment (leseb) [inline pkg/operator/ceph/cluster/cephstatus.go:153]: Why don't we return an error anymore?

File: pkg/operator/ceph/cluster/cluster.go
Comment (leseb) [inline pkg/operator/ceph/cluster/cluster.go:101]:
```suggestion
	opcontroller.UpdateCondition(c.context, c.namespacedName, cephv1.ConditionProgressing, v1.ConditionTrue, opcontroller.ClusterProgressingReason, "Reconciling mon(s)")
```

File: pkg/operator/ceph/cluster/cluster.go
Comment (leseb) [inline pkg/operator/ceph/cluster/cluster.go:145]:
```suggestion
	opcontroller.UpdateCondition(c.context, c.namespacedName, cephv1.ConditionProgressing, v1.ConditionTrue, opcontroller.ClusterProgressingReason, "Reconciling OSD(s)")
```

File: pkg/operator/ceph/cluster/cluster.go
Comment (leseb) [reply pkg/operator/ceph/cluster/cluster.go:101]: I wonder if "reconciling" is meaningful, it feels really attached to the controller/operator patter where I think we are trying to indicate progress in the CRD.

What about "Configuring Ceph $COMPONENT" instead? IMO it's clearer for users to understand what's going on.

**Review by leseb — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/cluster/cephstatus.go
Comment (travisn) [reply pkg/operator/ceph/cluster/cephstatus.go:153]: The callers were only logging the error. Updating the status has been treated as best effort, so the main reconcile won't be blocked by updates to the status. Anyway, the status will be updated at the next 60s when the status is queried again, even if a single call fails.

File: pkg/operator/ceph/cluster/cluster.go
Comment (travisn) [reply pkg/operator/ceph/cluster/cluster.go:101]: Agreed, "Configuring Ceph $COMPONENT" sounds better

File: pkg/operator/ceph/cluster/cluster.go
Comment (travisn) [reply pkg/operator/ceph/cluster/cluster.go:145]: All production clusters will have multiple mons and osds, so I don't think we need to make the (s) optional.

File: pkg/operator/ceph/controller/conditions.go
Comment (travisn) [reply pkg/operator/ceph/controller/conditions.go:38]: Idk, if there are only two or three, it keeps the code simpler with them defined inline below.

File: pkg/operator/ceph/controller/conditions.go
Comment (travisn) [reply pkg/operator/ceph/controller/conditions.go:36]: As much as I have dragged my feet in the past to use this pattern :), I'll go ahead and do this. I just like strings too much!

**Review by leseb — APPROVED** [review]

File: Documentation/ceph-cluster-crd.md
Comment (BlaineEXE) [inline Documentation/ceph-cluster-crd.md:674]: I think we should refer to how k8s documents its conditions to be closer to what users will expect. I don't really interpret the `reason` as being the actual condition. The condition is actually the `type` as implied by the k8s nodes condition docs, not the `reason` as our code and docs seem to treat it.

https://kubernetes.io/docs/concepts/architecture/nodes/#condition

IMO, we always need to have the `type: Ready` condition, but the status will be `"False"` until the cluster is created, and we must set it back to `"False"` if the cluster falls into a non-ready state. The `reason` for the cluster condition should be `ClusterCreated` or `ClusterConnected` depending on whether it is an external cluster or not.

Having said this, I wonder if we should have a persistent `type: ClusterProgressing` condition that can be true/false depending on if the  operator is running a reconcile or not.

Comment (travisn) [reply Documentation/ceph-cluster-crd.md:674]: > I think we should refer to how k8s documents its conditions to be closer to what users will expect. I don't really interpret the `reason` as being the actual condition. The condition is actually the `type` as implied by the k8s nodes condition docs, not the `reason` as our code and docs seem to treat it.
> 
> https://kubernetes.io/docs/concepts/architecture/nodes/#condition

There is a predictable mapping between the condition and the reason so they are in reality somewhat interchangeable, but agreed that the focus should be on the type. I'll update the doc. 

> IMO, we always need to have the `type: Ready` condition, but the status will be `"False"` until the cluster is created, and we must set it back to `"False"` if the cluster falls into a non-ready state. The `reason` for the cluster condition should be `ClusterCreated` or `ClusterConnected` depending on whether it is an external cluster or not.

Whether the cluster is actually ready is a Ceph question. Rather than attempt to reflect the Ceph readiness in the conditions, the approach is to defer to the Ceph health status if that question needs to be answered. So the `Ready` condition is just to indicate if the cluster has completed initial reconcile to indicate if it's expected the storage can now be consumed. We expect the cluster to always be in a consumable state after that point unless the Ceph health status is not ok, in which case we refer back to the Ceph status instead of the conditions. Rolling the Ceph status into the conditions is just too complex IMO. I believe consumers such as OCS are also relying on the `Ready` status to indicate that operations dependent on the basic reconcile can be continued.

> Having said this, I wonder if we should have a persistent `type: ClusterProgressing` condition that can be true/false depending on if the operator is running a reconcile or not. (Aside: `ClusterProgressing` might not be a great name either... `ClusterReconciling` could be a little clearer, though the message we put on the condition can help clear up any misconceptions if we want to keep that for backwards compatibility)

Having a `ClusterProgressing: false` was actually the issue that spurred this whole change. Any `false` condition was causing Rancher to interpret it as an error condition. Therefore, if the cluster is not progressing, we now remove that condition and `false` conditions are reserved for error conditions.

Comment (BlaineEXE) [reply Documentation/ceph-cluster-crd.md]: > Having a `ClusterProgressing: false` was actually the issue that spurred this whole change. Any `false` condition was causing Rancher to interpret it as an error condition. Therefore, if the cluster is not progressing, we now remove that condition and `false` conditions are reserved for error conditions.

Really!? But how does Rancher handle the Kubernetes node conditions like "DiskPressure" "False"?!?!

Comment (BlaineEXE) [reply Documentation/ceph-cluster-crd.md]: To further prove my point, the k8s sig-architecture api conventions do not suggest that non-True conditions should be treated as errors. Instead, they should have "abnormal-true" polarity, which suggests the opposite of what Rancher is doing.

https://github.com/kubernetes/community/blob/a2cdce5/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties

**Review by BlaineEXE — APPROVED** [review]
Approved after clarifying my understanding of Rancher's behavior to be really only concerned about the "Progressing" condition.

**Pull request merged**
